tests/Unit/API/Site/Controllers/MerchantCenter/AttributeMappingRulesControllerTest.php

```diff
@@ -0,0 +1,194 @@
+	 * @var AttributeMappingHelper
+	 */
+	private AttributeMappingHelper $attribute_mapping_helper;
+
+
+	/**
+	 * @var AttributeMappingRulesQuery
+	 */
+	private AttributeMappingRulesQuery $attribute_mapping_rules_query;
+
+	public function setUp(): void {
+		parent::setUp();
+		$this->attribute_mapping_helper      = $this->createMock( AttributeMappingHelper::class );
+		$this->attribute_mapping_rules_query = $this->createMock( AttributeMappingRulesQuery::class );
+		$this->controller                    = new AttributeMappingRulesController( $this->server, $this->attribute_mapping_helper, $this->attribute_mapping_rules_query );
+		$this->controller->register();
+	}
+
+
+	public function test_register_route() {
+		$this->assertArrayHasKey( self::ROUTE_RULES, $this->server->get_routes() );
+	}
+
+	public function test_get_rules_route() {
+		$data = [
+			[
+				'id'                      => 1,
+				'attribute'               => 'adult',
+				'source'                  => 'yes',
+				'category_condition_type' => 'ALL',
+				'categories'              => '',
+			],
+			[
+				'id'              => 2,
+				'attribute'       => 'brand',
+				'source'          => 'taxonomy:product_brand',
+				'categories_type' => 'ONLY',
+				'categories'      => '1,2,3',
+			],
+		];
+
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( 'get_results' )->willReturn( $data );
+
+		$response = $this->do_request( self::ROUTE_RULES );
+
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertEquals( $data, $response->get_data() );
+	}
+
+	public function test_create_rule_route() {
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( 'last_insert_id' )->willReturn( self::TEST_RULE_ID );
+
+		$this->validate_post_route( self::ROUTE_RULES, 'insert' );
+	}
+
+	public function test_update_rule_route() {
+		$this->validate_post_route( self::ROUTE_RULE, 'update' );
+	}
+
+	public function test_create_rule_route_fails() {
+		$this->attribute_mapping_helper->expects( $this->any() )
+			->method( 'get_attributes' )->willReturn(
+				self::TEST_ATTRIBUTES
+			);
+
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( 'insert' )->willReturn( 0 );
+
+		$response = $this->do_request( self::ROUTE_RULES, 'post', self::TEST_RULE );
+		$this->assertEquals( 400, $response->get_status() );
+	}
+
+	public function test_update_rule_route_fails() {
+		$this->attribute_mapping_helper->expects( $this->any() )
+			->method( 'get_attributes' )->willReturn(
+				self::TEST_ATTRIBUTES
+			);
+
+		$this->attribute_mapping_rules_query->expects( $this->once() )
+			->method( 'update' )->willReturn( 0 );
+
+		$response = $this->do_request( self::ROUTE_RULE, 'post', self::TEST_RULE );
+		$this->assertEquals( 400, $response->get_status() );
+	}
+
+	private function validate_post_route( $route, $method ) {
+		$rule_with_id = array_merge( [ 'id' => self::TEST_RULE_ID ], self::TEST_RULE, [ 'categories' => null ] );
```

Comment:
> This has been refactored, since id is marked as readonly it doesn't produce an error. But we shouldn't be sending the ID in the data to mimic what a real request would look like.

`rule_with_id` is used only for return values, not for the input being send. For the returned value yes we have the id.

Reply:
> Maybe we also don't need a shared validate function anymore. Since the create and update endpoints share the same schema we only need to check the validation rules once.

Wouldn't this make unsafe the controller tests? For example if we would delete the schema lines accidentally in the controller for one of the update/create routes.  The test will pass silently. Checking both we are sure that both of them are working.